### ISSUE TITLE
react/prop-types breaks when propTypes use spread operator

### ISSUE THREAD

**Issue by EricLambrecht** [yannickcr/eslint-plugin-react#2195]
In a component similar to this one:

```
static propTypes = {
  foo: PropTypes.string,
}

componentDidMount() {
  const { foo, bar } = this.props;
  console.log(foo, bar);
}
```

`react/prop-types` complains that `'bar' is missing in props validation`, **which is correct.**

However, if I change the propTypes to

```
static propTypes = {
  foo: PropTypes.string,
  ...{ other: PropTypes.bool, stuff: PropTypes.func },
}
```

eslint will pass, which **seems to be a bug.**

I did not find any similar issue here on github, but maybe I just missed it.

**Comment by ljharb** [comment]
The spread is basically impossible to statically analyze; but I’d still expect the error on foo, since it’s defined statically.

**Comment by EricLambrecht** [comment]
You mean `bar`, right?

I think the addition of _something_, even if not statically analyzable, should not prevent the rest from being statically analyzable, because it seems like this is the case.

**Comment by ljharb** [comment]
ahhh right, that's why. Because we can't determine if `bar` is present or not, we currently treat a spread as "all possible props".

If we errored on `bar` here, and you'd actually brought it in via the spread, then you'd be forced to silence the false positive, or to duplicate the `bar` declaration.

**Comment by EricLambrecht** [comment]
I see! That's what I presumed. Thank you for clarifying that!

I wonder if would make sense to make an option out of that:
1. throw an error regardless of the spread, which could lead to false positives
2. Treat the spread as "all possible prop" as it is right now.

I don't know if that's worth the effort, though.

**Comment by rohan-deshpande** [comment]
I'm noticing something similar. I have the following rules set up

```json
"react/prop-types": [
  "error",
  {
    "ignore": [],
    "customValidators": [],
    "skipUndeclared": false
  }
],
```

And the following component

```js
const Force = ({ type, emitterId, id, fx, fy, fz, life, easing, update }) => (
  <Behaviour {...{ type, emitterId, id, easing, update }}>
    <ControlsBlock>
      <NumberInput
        defaultValue={fx}
        label="fx"
        onValueUpdate={fx => update(type, emitterId, id, { fx })}
      />
      <NumberInput
        defaultValue={fy}
        label="fy"
        onValueUpdate={fy => update(type, emitterId, id, { fy })}
      />
      <NumberInput
        defaultValue={fz}
        label="fz"
        onValueUpdate={fz => update(type, emitterId, id, { fz })}
      />
    </ControlsBlock>
  </Behaviour>
);
```

Does not cause any prop type validations to display errors. If I remove the spread object completely, the rules work again.

**Comment by ljharb** [comment]
I’d suggest avoiding spreading props entirely; one big benefit is regaining the static analysis that doing so prevents.

**Comment by rohan-deshpande** [comment]
What you suggest is going against DRY quite a bit though and this is an extremely common React pattern so I don’t think people are going to stop doing it.

**Comment by brandonaaskov** [comment]
@ljharb thanks for the "[you're holding your phone wrong](https://www.engadget.com/2010-06-24-apple-responds-over-iphone-4-reception-issues-youre-holding-th.html)" kind of comment. Is there any technical reason that you're not supporting basics of the language, or is just your opinion pushed onto others?

**Comment by ljharb** [comment]
@brandonaaskov the technical reason is that it's impractical and/or impossible to figure out, statically, what's inside spread props. My opinion is that this problem shouldn't come up in the first place - but the problem remains unavoidable.

There's plenty basics in the language that nobody should ever use; that's what linters are *for* - to stop you from using bad parts of the language.

**Comment by brandonaaskov** [comment]
@ljharb are you saying that the spread operator is, in your opinion (I really need to keep emphasizing that word), a bad part of the language? I think a lot of people would disagree with you. I _think_ that. I don't _know_ that. Kind of like an opinion.

I get that this is open source software, but I didn't write it and open source it, y'all did. There's an expectation and assumption that the creators of any product actually care about the audience they're creating it for. So when you say that it's "impractical" I would say, to whom? Doesn't seem impractical to me as an end user of these tools.

And when you say that it's "impossible to figure out, statically..." you're again misunderstanding your audience. Most of the people using these tools are using them _because_ they don't want to know/care about that lower level stuff. Which makes projects like this great. But it makes responses like yours condescending and unhelpful. Since the problem remains unavoidable by your own admission, perhaps it's a problem worth solving.

I'm going to point out one last time that you're pushing your opinions and not listening. You said yourself:
"I’d suggest avoiding spreading props entirely; one big benefit is regaining the static analysis that doing so prevents." You'd _suggest_.... 

Well a suggestion is one thing, but now you're saying it's a must. And if I don't care about that "big benefit", I should be able to opt out of it. React, for example, suggests that you shouldn't set HTML from code with `innerHTML`, [and gives you the opportunity to ignore that suggestion](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml). They let you opt out of that "big benefit." 

And lastly, "There's plenty basics in the language that nobody should ever use; that's what linters are for".... Uhhhhh.... No??? Do you really think a linter's primary function is to "fix" broken aspects of the language? Maybe that's what you think they're for, but for the majority of javascript devs, the primary reason is to maintain a consistent code style across codebases and teams.

**Comment by ljharb** [comment]
@brandonaaskov no, i'm saying that "spread syntax exists in the language" is, like for every language feature, insufficient justification on its own to use it.

Yes, a linter's primary function is ABSOLUTELY to fix the broken parts of the language, including "there's more than one way to do something".

This issue has "help wanted". If you want it to happen, write a PR. Effectively accusing me of not improving the software because of my personal opinions is hostile, inappropriate, unprofessional, and unwelcome here.

**Comment by brandonaaskov** [comment]
@ljharb... you're just gonna shut down the discussion unceremoniously? Great communicator. What an adult. I didn't even say anything mean or offensive. I had no idea you had such delicate sensibilities.
![boo this man](https://user-images.githubusercontent.com/445778/125997926-4e4a10a1-5d08-4e87-a657-923dba4e8469.gif)

**Comment by brandonaaskov** [comment]
And for the record, I accused you of _not listening_ to your audience because your ego is in the way. And I stand by that, but I'm happy to leave that comment on Twitter or something if you think that's a better place for it.

**Comment by ljharb** [comment]
I marked the issue as "help wanted", long before offering my opinion. That means I have ALREADY listened, and explicitly indicated a PR to solve it would be accepted.

You're accusation is ridiculous, because offering an opinion is not "not listening".

**Comment by ljharb** [comment]
Given the options in https://github.com/yannickcr/eslint-plugin-react/issues/2195#issuecomment-472358331, I think there's nothing to be done here.

If someone believes differently, a PR with test cases would be welcome.

**Comment by kcjonson** [comment]
@ljharb can you explain why "The spread is basically impossible to statically analyze" I'm not seeing how that is true. 
If you mean within the same file, you're right, since variables could come from anywhere. It would require that the code tree walk to find the source of the variable that is being spread, that's possible but would have negative performance impact.

I still feel that this is critical to remaining DRY propType declarations.

**Comment by ljharb** [comment]
eslint only operates within the same file.
